services/121-service/test/performance/export-100k-registrations-ocw.test.ts

```
@@ -0,0 +1,72 @@
+
+const duplicateLowNumber = 5;
+const duplicateHighNumber = 17; // cronjob duplicate number should be 2^17 = 131072
+const testTimeout = 150_000; // 2,5 minutes
+const duplicateNumber =
+  // eslint-disable-next-line n/no-process-env -- Required to detect high data volume mode for performance testing
+  process.env.HIGH_DATA_VOLUME === 'true'
+    ? duplicateHighNumber
+    : duplicateLowNumber;
+
+jest.setTimeout(testTimeout);
+describe('Export High volume of registrations', () => {
+  let accessToken: string;
+
+  it('Setup, add registrations and export', async () => {
+    // Arrange
+    const startTime = Date.now();
+    await resetDB(SeedScript.nlrcMultiple, __filename);
+    accessToken = await getAccessToken();
+    // Upload registration
+    const importRegistrationResponse = await importRegistrations(
+      programIdOCW,
+      [registrationVisa],
+      accessToken,
+    );
+    expect(importRegistrationResponse.statusCode).toBe(HttpStatus.CREATED);
+
+    // Duplicate registration to be more than 100k
+    const duplicateRegistrationsResponse =
+      await duplicateRegistrationsAndPaymentData({
+        powerNumberRegistration: duplicateNumber,
+        numberOfPayments: 0,
+        accessToken,
+        body: {
+          secret: env.RESET_SECRET,
+        },
+      });
+    expect(duplicateRegistrationsResponse.statusCode).toBe(HttpStatus.CREATED);
+
+    // Export registrations
+    const exportResponse = await exportAllRegistrations(
+      programIdOCW,
+      accessToken,
+    );
+
+    expect(exportResponse.statusCode).toBe(HttpStatus.OK);
+
+    // Verify we exported a significant number of registrations
+    const expectedMinRegistrations =
+      duplicateNumber === duplicateHighNumber ? 100000 : 25;
+    expect(exportResponse.body.data.length).toBeGreaterThan(
+      expectedMinRegistrations,
+    );
+    const elapsedTime = Date.now() - startTime;
+    expect(elapsedTime).toBeLessThan(testTimeout);
```

`testTimeout` is playing 2 roles: it's
- used to time out the test so that it doesn't take too long
- we assert that the elapsedTime is less than `testTimeout`

But if the test would ever take longer than `testTimeout` we'd never reach this line. So either we should have 2 different variables (`testTimeout` and `maximumTimeAllowed` for example) because we want to use them for different things, or we _only_ use `testTimeout` for the actual timeout and let the test just fail if takes too long.

I vote for introducing another variable because we don't really care about how long the setup takes, just how long exporting the 100K registrations takes. We can then use `testTimeout` to make sure the test as a whole times out when it's taking way too long.

---

Changed